File: scripts/report_utils.py
#!/usr/bin/env python3
"""
Standardized report generation utilities for optics textbook scripts.

This module provides consistent report generation across all scripts,
with support for multiple output formats and standardized file locations.

Key Features:
    - Multi-format report generation (Markdown, JSON, plain text)
    - Centralized output directory (reports/)
    - Timestamp support for versioned reports
    - Fluent API for building markdown reports
    - Standardized validation and file list report templates
    - Automatic directory creation

Classes:
    ReportGenerator: Main class for generating reports in multiple formats
    MarkdownReportBuilder: Fluent API for building structured markdown reports

Utility Functions:
    create_validation_report: Generate standardized validation reports
    create_file_list_report: Generate simple file listing reports
    print_report_paths: Pretty-print generated report paths

Usage Examples:

    Basic report generation:
        >>> from report_utils import ReportGenerator
        >>>
        >>> gen = ReportGenerator("my_report")
        >>> gen.write_markdown("# Results\\n\\nAll checks passed.")
        >>> gen.write_json({"status": "success", "count": 42})
        >>> gen.write_text(["file1.txt", "file2.txt"])

    Building structured markdown:
        >>> from report_utils import MarkdownReportBuilder
        >>>
        >>> builder = MarkdownReportBuilder("Analysis Report")
        >>> builder.add_section("Findings")
        >>> builder.add_list(["Finding 1", "Finding 2"])
        >>> builder.add_table(["File", "Issues"], [
        ...     ["test.md", "3"],
        ...     ["example.md", "1"]
        ... ])
        >>> markdown = builder.build()
        >>>
        >>> gen = ReportGenerator("analysis")
        >>> gen.write_markdown(markdown)

    Validation reports:
        >>> from report_utils import create_validation_report, ReportGenerator
        >>>
        >>> issues = {
        ...     "missing_refs": ["ref1", "ref2"],
        ...     "broken_links": ["link1"]
        ... }
        >>> summary = {"total_issues": 3, "files_checked": 10}
        >>>
        >>> markdown, json_data = create_validation_report(
        ...     "Validation Results", issues, summary
        ... )
        >>>
        >>> gen = ReportGenerator("validation_report")
        >>> gen.write_all(markdown, json_data)

    File list reports:
        >>> from report_utils import create_file_list_report, ReportGenerator
        >>>
        >>> files = ["img1.png", "img2.png", "img3.png"]
        >>> markdown, text = create_file_list_report(
        ...     "Unreferenced Images",
        ...     files,
        ...     description="Images not referenced in any markdown file"
        ... )
        >>>
        >>> gen = ReportGenerator("unreferenced_images")
        >>> gen.write_markdown(markdown)
        >>> gen.write_text(text)

Integration:
    This module works seamlessly with shared_utils.py for validation and
    file operations. All reports are saved to the reports/ directory by default,
    which is automatically created if it doesn't exist.

    Reports can include timestamps for versioning:
        >>> gen = ReportGenerator("deletion_log")
        >>> gen.write_json(data, include_timestamp=True)
        # Creates: reports/deletion_log_20251006_143052.json

Author: Optics Textbook Scripts Package
Version: 1.0.0
"""

import json
from pathlib import Path
from datetime import datetime
from typing import Any, Dict, List, Optional, Union, Tuple


class ReportGenerator:
    """Generate standardized reports in various formats."""

    def __init__(self, report_name: str, output_dir: str = 'reports'):
        """
        Initialize report generator.

        Args:
            report_name: Base name for report files (without extension)
            output_dir: Directory to store reports (default: 'reports')
        """
        self.report_name = report_name
        self.output_dir = Path(output_dir)
        self.output_dir.mkdir(exist_ok=True)
        self.timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')

    def get_path(self, extension: str, include_timestamp: bool = False) -> Path:
        """
        Get full path for report file.

        Args:
            extension: File extension (without dot)
            include_timestamp: Whether to include timestamp in filename

        Returns:
            Path object for the report file
        """
        if include_timestamp:
            filename = f"{self.report_name}_{self.timestamp}.{extension}"
        else:
            filename = f"{self.report_name}.{extension}"
        return self.output_dir / filename

    def write_markdown(self, content: str, include_timestamp: bool = False) -> Path:
        """
        Write markdown report.

        Args:
            content: Markdown content to write
            include_timestamp: Whether to include timestamp in filename

        Returns:
            Path to written file
        """
        filepath = self.get_path('md', include_timestamp)
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        return filepath

    def write_json(self, data: Union[Dict, List], include_timestamp: bool = False) -> Path:
        """
        Write JSON report.

        Args:
            data: Data to serialize as JSON
            include_timestamp: Whether to include timestamp in filename

        Returns:
            Path to written file
        """
        filepath = self.get_path('json', include_timestamp)
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        return filepath

    def write_text(self, lines: Union[List[str], str], include_timestamp: bool = False) -> Path:
        """
        Write plain text report.

        Args:
            lines: List of lines or single string to write
            include_timestamp: Whether to include timestamp in filename

        Returns:
            Path to written file
        """
        filepath = self.get_path('txt', include_timestamp)
        content = lines if isinstance(lines, str) else '\n'.join(lines)
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        return filepath

    def write_all(self, markdown_content: str, json_data: Dict,
                  text_lines: Optional[List[str]] = None,
                  include_timestamp: bool = False) -> Dict[str, Path]:
        """
        Write report in all formats at once.

        Args:
            markdown_content: Markdown formatted content
            json_data: JSON serializable data
            text_lines: Optional plain text lines
            include_timestamp: Whether to include timestamp in filenames

        Returns:
            Dictionary mapping format to written file path
        """
        paths = {
            'markdown': self.write_markdown(markdown_content, include_timestamp),
            'json': self.write_json(json_data, include_timestamp)
        }
        if text_lines:
            paths['text'] = self.write_text(text_lines, include_timestamp)
        return paths


class MarkdownReportBuilder:
    """Helper class for building markdown reports."""

    def __init__(self, title: str):
        """
        Initialize markdown report builder.

        Args:
            title: Report title
        """
        self.lines = []
        self.add_title(title)
        self.add_metadata()

    def add_title(self, title: str, level: int = 1) -> 'MarkdownReportBuilder':
        """Add a title/heading."""
        self.lines.append(f"{'#' * level} {title}\n")
        return self

    def add_metadata(self) -> 'MarkdownReportBuilder':
        """Add report generation metadata."""
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        self.lines.append(f"*Generated: {timestamp}*\n")
        return self

    def add_section(self, title: str, level: int = 2) -> 'MarkdownReportBuilder':
        """Add a section heading."""
        self.lines.append(f"\n{'#' * level} {title}\n")
        return self

    def add_text(self, text: str) -> 'MarkdownReportBuilder':
        """Add plain text paragraph."""
        self.lines.append(f"{text}\n")
        return self

    def add_list(self, items: List[str], ordered: bool = False) -> 'MarkdownReportBuilder':
        """Add a list (ordered or unordered)."""
        for i, item in enumerate(items, 1):
            prefix = f"{i}." if ordered else "-"
            self.lines.append(f"{prefix} {item}\n")
        return self

    def add_table(self, headers: List[str], rows: List[List[str]]) -> 'MarkdownReportBuilder':
        """Add a markdown table."""
        # Headers
        self.lines.append("| " + " | ".join(headers) + " |\n")
        # Separator
        self.lines.append("| " + " | ".join(["---"] * len(headers)) + " |\n")
        # Rows
        for row in rows:
            self.lines.append("| " + " | ".join(str(cell) for cell in row) + " |\n")
        self.lines.append("\n")
        return self

    def add_code_block(self, code: str, language: str = "") -> 'MarkdownReportBuilder':
        """Add a code block."""
        self.lines.append(f"```{language}\n{code}\n```\n")
        return self

    def add_summary(self, summary_dict: Dict[str, Any]) -> 'MarkdownReportBuilder':
        """Add a summary section with key-value pairs."""
        self.add_section("Summary", level=2)
        for key, value in summary_dict.items():
            self.lines.append(f"- **{key}**: {value}\n")
        return self

    def build(self) -> str:
        """Build and return the complete markdown document."""
        return ''.join(self.lines)


def create_validation_report(title: str, issues: Dict[str, List[Any]],
                             summary: Dict[str, int]) -> Tuple[str, Dict[str, Any]]:
    """
    Create a standardized validation report.

    Args:
        title: Report title
        issues: Dictionary of issue categories to issue lists
        summary: Summary statistics

    Returns:
        Tuple of (markdown_content, json_data)
    """
    builder = MarkdownReportBuilder(title)

    # Add summary
    builder.add_summary(summary)

    # Add issues by category
    for category, issue_list in issues.items():
        if issue_list:
            builder.add_section(f"{category.replace('_', ' ').title()}", level=2)
            builder.add_text(f"Found {len(issue_list)} issue(s):\n")
            builder.add_list([str(issue) for issue in issue_list])

    # Build JSON data
    json_data = {
        'title': title,
        'timestamp': datetime.now().isoformat(),
        'summary': summary,
        'issues': issues
    }

    return builder.build(), json_data


def create_file_list_report(title: str, files: List[str],
                            description: str = "") -> Tuple[str, str]:
    """
    Create a simple file list report.

    Args:
        title: Report title
        files: List of file paths
        description: Optional description

    Returns:
        Tuple of (markdown_content, text_content)
    """
    builder = MarkdownReportBuilder(title)

    if description:
        builder.add_text(description)

    builder.add_section("Files", level=2)
    builder.add_text(f"Total: {len(files)} files\n")
    builder.add_list(files)

    markdown = builder.build()
    text = '\n'.join(files)

    return markdown, text


def print_report_paths(paths: Dict[str, Path]) -> None:
    """
    Print report file paths in a consistent format.

    Args:
        paths: Dictionary mapping format to file path
    """
    print("\nReports generated:")
    for format_name, path in paths.items():
        print(f"  - {format_name.capitalize()}: {path}")

File: scripts/report-utils.js
#!/usr/bin/env node
/**
 * Standardized report generation utilities for MyST markdown project scripts.
 *
 * This module provides consistent report generation across all scripts,
 * with support for multiple output formats and standardized file locations.
 *
 * Key Features:
 *     - Multi-format report generation (Markdown, JSON, plain text)
 *     - Centralized output directory (reports/)
 *     - Timestamp support for versioned reports
 *     - Fluent API for building markdown reports
 *     - Standardized validation and file list report templates
 *     - Automatic directory creation
 *
 * Classes:
 *     ReportGenerator: Main class for generating reports in multiple formats
 *     MarkdownReportBuilder: Fluent API for building structured markdown reports
 *
 * Utility Functions:
 *     createValidationReport: Generate standardized validation reports
 *     createFileListReport: Generate simple file listing reports
 *     printReportPaths: Pretty-print generated report paths
 *
 * Author: MyST Project Scripts Package
 * Version: 1.0.0
 */

const fs = require('fs');
const path = require('path');

/**
 * Get current timestamp in YYYYMMDD_HHMMSS format
 */
function getTimestamp() {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const day = String(now.getDate()).padStart(2, '0');
  const hours = String(now.getHours()).padStart(2, '0');
  const minutes = String(now.getMinutes()).padStart(2, '0');
  const seconds = String(now.getSeconds()).padStart(2, '0');
  return `${year}${month}${day}_${hours}${minutes}${seconds}`;
}

/**
 * Get current ISO timestamp
 */
function getISOTimestamp() {
  return new Date().toISOString();
}

/**
 * Get formatted timestamp for display
 */
function getFormattedTimestamp() {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const day = String(now.getDate()).padStart(2, '0');
  const hours = String(now.getHours()).padStart(2, '0');
  const minutes = String(now.getMinutes()).padStart(2, '0');
  const seconds = String(now.getSeconds()).padStart(2, '0');
  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}

/**
 * Generate standardized reports in various formats
 */
class ReportGenerator {
  /**
   * Initialize report generator.
   *
   * @param {string} reportName - Base name for report files (without extension)
   * @param {string} outputDir - Directory to store reports (default: 'reports')
   */
  constructor(reportName, outputDir = 'reports') {
    this.reportName = reportName;
    this.outputDir = path.resolve(outputDir);
    this.timestamp = getTimestamp();

    // Create output directory if it doesn't exist
    if (!fs.existsSync(this.outputDir)) {
      fs.mkdirSync(this.outputDir, { recursive: true });
    }
  }

  /**
   * Get full path for report file.
   *
   * @param {string} extension - File extension (without dot)
   * @param {boolean} includeTimestamp - Whether to include timestamp in filename
   * @returns {string} Full path for the report file
   */
  getPath(extension, includeTimestamp = false) {
    let filename;
    if (includeTimestamp) {
      filename = `${this.reportName}_${this.timestamp}.${extension}`;
    } else {
      filename = `${this.reportName}.${extension}`;
    }
    return path.join(this.outputDir, filename);
  }

  /**
   * Write markdown report.
   *
   * @param {string} content - Markdown content to write
   * @param {boolean} includeTimestamp - Whether to include timestamp in filename
   * @returns {string} Path to written file
   */
  writeMarkdown(content, includeTimestamp = false) {
    const filepath = this.getPath('md', includeTimestamp);
    fs.writeFileSync(filepath, content, 'utf8');
    return filepath;
  }

  /**
   * Write JSON report.
   *
   * @param {Object|Array} data - Data to serialize as JSON
   * @param {boolean} includeTimestamp - Whether to include timestamp in filename
   * @returns {string} Path to written file
   */
  writeJson(data, includeTimestamp = false) {
    const filepath = this.getPath('json', includeTimestamp);
    fs.writeFileSync(filepath, JSON.stringify(data, null, 2), 'utf8');
    return filepath;
  }

  /**
   * Write plain text report.
   *
   * @param {Array<string>|string} lines - List of lines or single string to write
   * @param {boolean} includeTimestamp - Whether to include timestamp in filename
   * @returns {string} Path to written file
   */
  writeText(lines, includeTimestamp = false) {
    const filepath = this.getPath('txt', includeTimestamp);
    const content = Array.isArray(lines) ? lines.join('\n') : lines;
    fs.writeFileSync(filepath, content, 'utf8');
    return filepath;
  }

  /**
   * Write report in all formats at once.
   *
   * @param {string} markdownContent - Markdown formatted content
   * @param {Object} jsonData - JSON serializable data
   * @param {Array<string>|null} textLines - Optional plain text lines
   * @param {boolean} includeTimestamp - Whether to include timestamp in filenames
   * @returns {Object} Dictionary mapping format to written file path
   */
  writeAll(markdownContent, jsonData, textLines = null, includeTimestamp = false) {
    const paths = {
      markdown: this.writeMarkdown(markdownContent, includeTimestamp),
      json: this.writeJson(jsonData, includeTimestamp)
    };
    if (textLines) {
      paths.text = this.writeText(textLines, includeTimestamp);
    }
    return paths;
  }
}

/**
 * Helper class for building markdown reports with fluent API
 */
class MarkdownReportBuilder {
  /**
   * Initialize markdown report builder.
   *
   * @param {string} title - Report title
   */
  constructor(title) {
    this.lines = [];
    this.addTitle(title);
    this.addMetadata();
  }

  /**
   * Add a title/heading
   *
   * @param {string} title - Title text
   * @param {number} level - Heading level (1-6)
   * @returns {MarkdownReportBuilder} this for chaining
   */
  addTitle(title, level = 1) {
    this.lines.push('#'.repeat(level) + ' ' + title + '\n');
    return this;
  }

  /**
   * Add report generation metadata
   *
   * @returns {MarkdownReportBuilder} this for chaining
   */
  addMetadata() {
    const timestamp = getFormattedTimestamp();
    this.lines.push(`*Generated: ${timestamp}*\n`);
    return this;
  }

  /**
   * Add a section heading
   *
   * @param {string} title - Section title
   * @param {number} level - Heading level (default: 2)
   * @returns {MarkdownReportBuilder} this for chaining
   */
  addSection(title, level = 2) {
    this.lines.push('\n' + '#'.repeat(level) + ' ' + title + '\n');
    return this;
  }

  /**
   * Add plain text paragraph
   *
   * @param {string} text - Text content
   * @returns {MarkdownReportBuilder} this for chaining
   */
  addText(text) {
    this.lines.push(text + '\n');
    return this;
  }

  /**
   * Add a list (ordered or unordered)
   *
   * @param {Array<string>} items - List items
   * @param {boolean} ordered - Whether to create an ordered list
   * @returns {MarkdownReportBuilder} this for chaining
   */
  addList(items, ordered = false) {
    items.forEach((item, index) => {
      const prefix = ordered ? `${index + 1}.` : '-';
      this.lines.push(`${prefix} ${item}\n`);
    });
    return this;
  }

  /**
   * Add a markdown table
   *
   * @param {Array<string>} headers - Table headers
   * @param {Array<Array<string>>} rows - Table rows
   * @returns {MarkdownReportBuilder} this for chaining
   */
  addTable(headers, rows) {
    // Headers
    this.lines.push('| ' + headers.join(' | ') + ' |\n');
    // Separator
    this.lines.push('| ' + headers.map(() => '---').join(' | ') + ' |\n');
    // Rows
    rows.forEach(row => {
      this.lines.push('| ' + row.map(cell => String(cell)).join(' | ') + ' |\n');
    });
    this.lines.push('\n');
    return this;
  }

  /**
   * Add a code block
   *
   * @param {string} code - Code content
   * @param {string} language - Language for syntax highlighting
   * @returns {MarkdownReportBuilder} this for chaining
   */
  addCodeBlock(code, language = '') {
    this.lines.push('```' + language + '\n' + code + '\n```\n');
    return this;
  }

  /**
   * Add a summary section with key-value pairs
   *
   * @param {Object} summaryDict - Dictionary of summary items
   * @returns {MarkdownReportBuilder} this for chaining
   */
  addSummary(summaryDict) {
    this.addSection('Summary', 2);
    Object.entries(summaryDict).forEach(([key, value]) => {
      this.lines.push(`- **${key}**: ${value}\n`);
    });
    return this;
  }

  /**
   * Build and return the complete markdown document
   *
   * @returns {string} Complete markdown content
   */
  build() {
    return this.lines.join('');
  }
}

/**
 * Create a standardized validation report.
 *
 * @param {string} title - Report title
 * @param {Object} issues - Dictionary of issue categories to issue lists
 * @param {Object} summary - Summary statistics
 * @returns {Array} [markdownContent, jsonData]
 */
function createValidationReport(title, issues, summary) {
  const builder = new MarkdownReportBuilder(title);

  // Add summary
  builder.addSummary(summary);

  // Add issues by category
  Object.entries(issues).forEach(([category, issueList]) => {
    if (issueList && issueList.length > 0) {
      const categoryTitle = category.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
      builder.addSection(categoryTitle, 2);
      builder.addText(`Found ${issueList.length} issue(s):\n`);
      builder.addList(issueList.map(issue => String(issue)));
    }
  });

  // Build JSON data
  const jsonData = {
    title,
    timestamp: getISOTimestamp(),
    summary,
    issues
  };

  return [builder.build(), jsonData];
}

/**
 * Create a simple file list report.
 *
 * @param {string} title - Report title
 * @param {Array<string>} files - List of file paths
 * @param {string} description - Optional description
 * @returns {Array} [markdownContent, textContent]
 */
function createFileListReport(title, files, description = '') {
  const builder = new MarkdownReportBuilder(title);

  if (description) {
    builder.addText(description);
  }

  builder.addSection('Files', 2);
  builder.addText(`Total: ${files.length} files\n`);
  builder.addList(files);

  const markdown = builder.build();
  const text = files.join('\n');

  return [markdown, text];
}

/**
 * Print report file paths in a consistent format.
 *
 * @param {Object} paths - Dictionary mapping format to file path
 */
function printReportPaths(paths) {
  console.log('\nReports generated:');
  Object.entries(paths).forEach(([formatName, filepath]) => {
    const capitalizedFormat = formatName.charAt(0).toUpperCase() + formatName.slice(1);
    console.log(`  - ${capitalizedFormat}: ${filepath}`);
  });
}

// ============================================================================
// Module Exports
// ============================================================================

module.exports = {
  ReportGenerator,
  MarkdownReportBuilder,
  createValidationReport,
  createFileListReport,
  printReportPaths,
  getTimestamp,
  getISOTimestamp,
  getFormattedTimestamp
};
